File: Dart/15-Greedy/Algorithms/1-HuffmanCoding/HuffmanCoding.dart
import 'dart:collection';
import 'package:collection/collection.dart'; // Import the collection package

// --- 1. Huffman Coding Using a Priority Queue (Min Heap) ---
class HuffmanNode {
  int frequency;
  String? character;
  HuffmanNode? left, right;

  HuffmanNode(this.frequency, {this.character, this.left, this.right});

  // Define comparison for the PriorityQueue
  int compareTo(HuffmanNode other) {
    return frequency.compareTo(other.frequency);
  }
}

Map<String, String> buildHuffmanCodes(HuffmanNode? node, String code, Map<String, String> codes) {
  if (node == null) return codes;

  if (node.character != null) {
    codes[node.character!] = code;
    return codes;
  }

  buildHuffmanCodes(node.left, code + '0', codes);
  buildHuffmanCodes(node.right, code + '1', codes);
  return codes;
}

String huffmanEncode(String text) {
  if (text.isEmpty) return "";

  // 1. Calculate frequency of each character
  Map<String, int> frequencyMap = {};
  for (int i = 0; i < text.length; i++) {
    String char = text[i];
    frequencyMap[char] = (frequencyMap[char] ?? 0) + 1;
  }

  // 2. Create a priority queue (min heap)
  PriorityQueue<HuffmanNode> pq = PriorityQueue<HuffmanNode>((a, b) => a.compareTo(b));
  frequencyMap.forEach((char, freq) {
    pq.add(HuffmanNode(freq, character: char));
  });

  // 3. Build the Huffman tree
  while (pq.length > 1) {
    HuffmanNode left = pq.removeFirst();
    HuffmanNode right = pq.removeFirst();
    HuffmanNode parent = HuffmanNode(left.frequency + right.frequency, left: left, right: right);
    pq.add(parent);
  }

  // 4. Get the Huffman codes
  HuffmanNode? root = pq.first;
  Map<String, String> huffmanCodes = {};
  if(root != null) {
    huffmanCodes = buildHuffmanCodes(root, '', huffmanCodes);
  }


  // 5. Encode the text
  String encodedText = '';
  for (int i = 0; i < text.length; i++) {
    encodedText += huffmanCodes[text[i]] ?? ''; //handle null huffmanCodes[text[i]]
  }
  return encodedText;
}

String huffmanDecode(String encodedText, Map<String, String> huffmanCodes) {
  if (encodedText.isEmpty) return "";

  // Reverse the huffman codes for easier decoding
  Map<String, String> reversedHuffmanCodes = {};
  huffmanCodes.forEach((char, code) {
    reversedHuffmanCodes[code] = char;
  });

  String decodedText = '';
  String currentCode = '';
  for (int i = 0; i < encodedText.length; i++) {
    currentCode += encodedText[i];
    if (reversedHuffmanCodes.containsKey(currentCode)) {
      decodedText += reversedHuffmanCodes[currentCode]!;
      currentCode = '';
    }
  }
  return decodedText;
}

void main() {
  // Example usage of Huffman Coding
  String text = "hello world";
  String encodedText = huffmanEncode(text);
  Map<String, String> huffmanCodes = {};
  //build huffman codes.
  if (text.isNotEmpty) {
    Map<String, int> frequencyMap = {};
    for (int i = 0; i < text.length; i++) {
      String char = text[i];
      frequencyMap[char] = (frequencyMap[char] ?? 0) + 1;
    }
    PriorityQueue<HuffmanNode> pq = PriorityQueue<HuffmanNode>((a, b) => a.compareTo(b));
    frequencyMap.forEach((char, freq) {
      pq.add(HuffmanNode(freq, character: char));
    });
    while (pq.length > 1) {
      HuffmanNode left = pq.removeFirst();
      HuffmanNode right = pq.removeFirst();
      HuffmanNode parent = HuffmanNode(left.frequency + right.frequency, left: left, right: right);
      pq.add(parent);
    }
    HuffmanNode? root = pq.first;
    if(root != null){
      huffmanCodes = buildHuffmanCodes(root, '', huffmanCodes);
    }

  }

  print("Original text: $text");
  print("Encoded text: $encodedText");
  print("Huffman Codes: $huffmanCodes");
  String decodedText = huffmanDecode(encodedText, huffmanCodes);
  print("Decoded text: $decodedText");
  print("--------------------------------------------------------");

  // --- 2. Huffman Coding Using a Sorted List ---
  String text2 = "banana";
  var result2 = huffmanCodingSortedList(text2);
  print("Huffman Coding with Sorted List:");
  print("Original text: $text2");
  print("Encoded text: ${result2['encodedText']}");
  print("Huffman Codes: ${result2['codes']}");
  print("Decoded text: ${huffmanDecode(result2['encodedText'], result2['codes'])}");
  print("--------------------------------------------------------");

  // --- 3. Huffman Coding Using Recursive Approach ---
  String text3 = "mississippi";
  var result3 = huffmanCodingRecursive(text3);
  print("Huffman Coding with Recursion:");
  print("Original text: $text3");
  print("Encoded text: ${result3['encodedText']}");
  print("Huffman Codes: ${result3['codes']}");
  print("Decoded text: ${huffmanDecode(result3['encodedText'], result3['codes'])}");
  print("--------------------------------------------------------");

  // --- 4. Huffman Coding with Optimized Frequency Counting ---
  String text4 = "aabbccddeeff";
  var result4 = huffmanCodingOptimized(text4);
  print("Huffman Coding with Optimized Frequency Counting:");
  print("Original text: $text4");
  print("Encoded text: ${result4['encodedText']}");
  print("Huffman Codes: ${result4['codes']}");
  print("Decoded text: ${huffmanDecode(result4['encodedText'], result4['codes'])}");
  print("--------------------------------------------------------");

  // --- 5. Huffman Coding with Prebuilt Frequency Table ---
  String text5 = "foobar";
  Map<String, int> frequencyTable = {
    'f': 1,
    'o': 2,
    'b': 1,
    'a': 1,
    'r': 1,
  };
  var result5 = huffmanCodingPrebuiltTable(text5, frequencyTable);
  print("Huffman Coding with Prebuilt Frequency Table:");
  print("Original text: $text5");
  print("Encoded text: ${result5['encodedText']}");
  print("Huffman Codes: ${result5['codes']}");
  print("Decoded text: ${huffmanDecode(result5['encodedText'], result5['codes'])}");
}

// --- 2. Huffman Coding Using a Sorted List ---
// Node class for the Huffman tree
class Node {
  int frequency;
  String? character;
  Node? left;
  Node? right;

  Node(this.frequency, {this.character, this.left, this.right});
}

// Function to build the Huffman tree
Node buildHuffmanTree(List<Node> nodes) {
  while (nodes.length > 1) {
    nodes.sort((a, b) => a.frequency.compareTo(b.frequency)); // Sort by frequency
    Node left = nodes.removeAt(0);
    Node right = nodes.removeAt(0);
    Node parent = Node(left.frequency + right.frequency, left: left, right: right);
    nodes.insert(0, parent); // Insert the new node back into the list
  }
  return nodes[0]; // Return the root of the tree
}

// Function to generate Huffman codes
Map<String, String> generateHuffmanCodes(Node? node, String code, Map<String, String> codes) {
  if (node == null) return codes;
  if (node.character != null) {
    codes[node.character!] = code;
    return codes;
  }
  generateHuffmanCodes(node.left, code + '0', codes);
  generateHuffmanCodes(node.right, code + '1', codes);
  return codes;
}

// Huffman coding function
Map<String, dynamic> huffmanCodingSortedList(String text) {
  // Calculate character frequencies
  Map<String, int> frequencyMap = {};
  for (int i = 0; i < text.length; i++) {
    String char = text[i];
    frequencyMap[char] = (frequencyMap[char] ?? 0) + 1;
  }

  // Create a list of nodes
  List<Node> nodes = frequencyMap.entries.map((entry) => Node(entry.value, character: entry.key)).toList();

  // Build the Huffman tree
  Node root = buildHuffmanTree(nodes);

  // Generate Huffman codes
  Map<String, String> codes = {};
  if(root != null){
    codes = generateHuffmanCodes(root, '', codes);
  }


  // Encode the text
  String encodedText = text.split('').map((char) => codes[char] ?? '').join(''); //handle null codes[char]

  return {'encodedText': encodedText, 'codes': codes};
}

// --- 3. Huffman Coding Using Recursive Approach ---
class RecursiveNode {
  int frequency;
  String? character;
  RecursiveNode? left, right;

  RecursiveNode(this.frequency, {this.character, this.left, this.right});
}

// Function to build Huffman tree recursively
RecursiveNode buildHuffmanTreeRecursive(Map<String, int> frequencyMap) {
  if (frequencyMap.isEmpty) {
    return RecursiveNode(0);
  }
  if (frequencyMap.length == 1) {
    String char = frequencyMap.keys.first;
    return RecursiveNode(frequencyMap[char]!, character: char);
  }

  // Find the two smallest frequencies
  List<MapEntry<String, int>> sortedFrequencies = frequencyMap.entries.toList()..sort((a, b) => a.value.compareTo(b.value));
  String char1 = sortedFrequencies[0].key;
  String char2 = sortedFrequencies[1].key;
  int freq1 = sortedFrequencies[0].value;
  int freq2 = sortedFrequencies[1].value;

  // Remove the two smallest frequencies from the map
  frequencyMap.remove(char1);
  frequencyMap.remove(char2);

  // Create a new parent node
  RecursiveNode parent = RecursiveNode(freq1 + freq2,
      left: RecursiveNode(freq1, character: char1),
      right: RecursiveNode(freq2, character: char2));

  // Recursively build the rest of the tree
  frequencyMap[char1 + char2] = parent.frequency; // Use a combined key
  return buildHuffmanTreeRecursive(frequencyMap);
}

// Function to generate Huffman codes
Map<String, String> generateHuffmanCodesRecursive(RecursiveNode? node, String code, Map<String, String> codes) {
  if (node == null) return codes;

  if (node.character != null) {
    codes[node.character!] = code;
    return codes;
  }

  generateHuffmanCodesRecursive(node.left, code + '0', codes);
  generateHuffmanCodesRecursive(node.right, code + '1', codes);
  return codes;
}

// Huffman coding function
Map<String, dynamic> huffmanCodingRecursive(String text) {
  // Calculate character frequencies
  Map<String, int> frequencyMap = {};
  for (int i = 0; i < text.length; i++) {
    String char = text[i];
    frequencyMap[char] = (frequencyMap[char] ?? 0) + 1;
  }

  // Build the Huffman tree
  RecursiveNode root = buildHuffmanTreeRecursive(frequencyMap);

  // Generate Huffman codes
  Map<String, String> codes = {};
  if(root != null){
    codes = generateHuffmanCodesRecursive(root, '', codes);
  }


  // Encode the text
  String encodedText = text.split('').map((char) => codes[char] ?? '').join(''); //handle null codes[char]


  return {'encodedText': encodedText, 'codes': codes};
}

// --- 4. Huffman Coding with Optimized Frequency Counting ---
Map<String, dynamic> huffmanCodingOptimized(String text) {
  // 1. Calculate frequency of each character using a list
  List<int> frequencies = List.filled(256, 0); // Assuming ASCII characters
  for (int i = 0; i < text.length; i++) {
    frequencies[text.codeUnitAt(i)]++;
  }

  // 2. Create a priority queue (min heap)
  PriorityQueue<HuffmanNode> pq = PriorityQueue<HuffmanNode>((a, b) => a.compareTo(b));
  for (int i = 0; i < 256; i++) {
    if (frequencies[i] > 0) {
      pq.add(HuffmanNode(frequencies[i], character: String.fromCharCode(i)));
    }
  }

  // 3. Build the Huffman tree
  while (pq.length > 1) {
    HuffmanNode left = pq.removeFirst();
    HuffmanNode right = pq.removeFirst();
    HuffmanNode parent = HuffmanNode(left.frequency + right.frequency, left: left, right: right);
    pq.add(parent);
  }

  // 4. Get the Huffman codes
  HuffmanNode? root = pq.first;
  Map<String, String> huffmanCodes = {};
  if(root != null) {
    huffmanCodes = buildHuffmanCodes(root, '', huffmanCodes);
  }

  // 5. Encode the text
  String encodedText = '';
  for (int i = 0; i < text.length; i++) {
    encodedText += huffmanCodes[text[i]] ?? '';  //handle null huffmanCodes[text[i]]
  }
  return {'encodedText': encodedText, 'codes': huffmanCodes};
}

// --- 5. Huffman Coding with Prebuilt Frequency Table ---
Map<String, dynamic> huffmanCodingPrebuiltTable(String text, Map<String, int> frequencyTable) {
  // 1. Create a priority queue (min heap)
  PriorityQueue<HuffmanNode> pq = PriorityQueue<HuffmanNode>((a, b) => a.compareTo(b));
  frequencyTable.forEach((char, freq) {
    pq.add(HuffmanNode(freq, character: char));
  });

  // 2. Build the Huffman tree
  while (pq.length > 1) {
    HuffmanNode left = pq.removeFirst();
    HuffmanNode right = pq.removeFirst();
    HuffmanNode parent = HuffmanNode(left.frequency + right.frequency, left: left, right: right);
    pq.add(parent);
  }

  // 3. Get the Huffman codes
  HuffmanNode? root = pq.first;
  Map<String, String> huffmanCodes = {};
  if(root != null){
    huffmanCodes = buildHuffmanCodes(root, '', huffmanCodes);
  }

  // 4. Encode the text
  String encodedText = '';
  for (int i = 0; i < text.length; i++) {
    encodedText += huffmanCodes[text[i]] ?? ''; //handle null huffmanCodes[text[i]]
  }
  return {'encodedText': encodedText, 'codes': huffmanCodes};
}

File: C++/15-Greedy/Algorithms/1-HuffmanCoding/HuffmanCoding.js
// Huffman Coding in JavaScript - 5 Approaches

// 1. Basic Huffman Coding (Priority Queue with Array)
// ------------------------------------------------------

/**
 * Node class for the Huffman tree.
 */
class HuffmanNode {
  constructor(char, freq) {
    this.char = char;
    this.freq = freq;
    this.left = null;
    this.right = null;
  }
}

/**
 * Builds the Huffman tree and returns the root node.
 * Uses a simple array as a priority queue (not very efficient for large inputs).
 */
function buildHuffmanTreeBasic(charFreq) {
  if (!charFreq || charFreq.length === 0) {
    return null; // Handle empty input
  }
  const nodes = charFreq.map(([char, freq]) => new HuffmanNode(char, freq));

  while (nodes.length > 1) {
    // Sort to get the two smallest frequencies
    nodes.sort((a, b) => a.freq - b.freq);
    const left = nodes.shift();
    const right = nodes.shift();
    const newNode = new HuffmanNode(null, left.freq + right.freq);
    newNode.left = left;
    newNode.right = right;
    nodes.push(newNode);
  }

  return nodes[0]; // Root of the Huffman tree
}

/**
 * Generates the Huffman codes from the Huffman tree.
 */
function generateHuffmanCodesBasic(node, code = "", codes = {}) {
  if (!node) return codes; //handle empty node
  if (node.char) {
    codes[node.char] = code;
    return codes;
  }

  generateHuffmanCodesBasic(node.left, code + "0", codes);
  generateHuffmanCodesBasic(node.right, code + "1", codes);
  return codes;
}

/**
 * Encodes the text using the Huffman codes.
 */
function encodeTextBasic(text, codes) {
  if (!codes) return ""; // Handle empty codes
  let encodedText = "";
  for (const char of text) {
    encodedText += codes[char] || ''; // Handle chars not in codes
  }
  return encodedText;
}

/**
 * Decodes the text using the Huffman tree.
 */
function decodeTextBasic(encodedText, root) {
  if (!root) return ""; // Handle empty tree
  let decodedText = "";
  let currentNode = root;
  for (const bit of encodedText) {
    if (bit === "0") {
      currentNode = currentNode.left;
    } else {
      currentNode = currentNode.right;
    }

    if (currentNode && currentNode.char) { //check if currentNode is valid
      decodedText += currentNode.char;
      currentNode = root; // Reset to the root for the next character
    }
    else if (!currentNode) {
      return decodedText; //handle invalid encoded text
    }
  }
  return decodedText;
}

// 2. Huffman Coding with Optimized Priority Queue (Min Heap)
// ------------------------------------------------------------
class MinHeap {
  constructor() {
    this.heap = [];
  }

  getLeftChildIndex(index) { return 2 * index + 1; }
  getRightChildIndex(index) { return 2 * index + 2; }
  getParentIndex(index) { return Math.floor((index - 1) / 2); }

  hasLeftChild(index) { return this.getLeftChildIndex(index) < this.heap.length; }
  hasRightChild(index) { return this.getRightChildIndex(index) < this.heap.length; }
  hasParent(index) { return this.getParentIndex(index) >= 0; }

  leftChild(index) { return this.heap[this.getLeftChildIndex(index)]; }
  rightChild(index) { return this.heap[this.getRightChildIndex(index)]; }
  parent(index) { return this.heap[this.getParentIndex(index)]; }

  size() { return this.heap.length; }

  peek() {
    if (this.size() === 0) return null;
    return this.heap[0];
  }

  poll() {
    if (this.size() === 0) return null;
    if (this.size() === 1) return this.heap.pop();

    const item = this.heap[0];
    this.heap[0] = this.heap.pop();
    this.heapifyDown(0);
    return item;
  }

  add(item) {
    this.heap.push(item);
    this.heapifyUp(this.heap.length - 1);
  }

  heapifyUp(index) {
    while (this.hasParent(index) && this.parent(index).freq > this.heap[index].freq) {
      this.swap(index, this.getParentIndex(index));
      index = this.getParentIndex(index);
    }
  }

  heapifyDown(index) {
    while (this.hasLeftChild(index)) {
      let smallerChildIndex = this.getLeftChildIndex(index);
      if (this.hasRightChild(index) && this.rightChild(index).freq < this.leftChild(index).freq) {
        smallerChildIndex = this.getRightChildIndex(index);
      }

      if (this.heap[index].freq < this.heap[smallerChildIndex].freq) {
        break;
      } else {
        this.swap(index, smallerChildIndex);
      }
      index = smallerChildIndex;
    }
  }

  swap(index1, index2) {
    const temp = this.heap[index1];
    this.heap[index1] = this.heap[index2];
    this.heap[index2] = temp;
  }
}

function buildHuffmanTreeOptimized(charFreq) {
  if (!charFreq || charFreq.length === 0) {
    return null;
  }
  const minHeap = new MinHeap();
  charFreq.forEach(([char, freq]) => {
    minHeap.add(new HuffmanNode(char, freq));
  });

  while (minHeap.size() > 1) {
    const left = minHeap.poll();
    const right = minHeap.poll();
    const newNode = new HuffmanNode(null, left.freq + right.freq);
    newNode.left = left;
    newNode.right = right;
    minHeap.add(newNode);
  }
  return minHeap.peek();
}

function generateHuffmanCodesOptimized(node, code = "", codes = {}) {
  if (!node) return codes;
  if (node.char) {
    codes[node.char] = code;
    return codes;
  }
  generateHuffmanCodesOptimized(node.left, code + "0", codes);
  generateHuffmanCodesOptimized(node.right, code + "1", codes);
  return codes;
}

function encodeTextOptimized(text, codes) {
  if (!codes) return "";
  let encodedText = "";
  for (const char of text) {
    encodedText += codes[char] || '';
  }
  return encodedText;
}

function decodeTextOptimized(encodedText, root) {
  if (!root) return "";
  let decodedText = "";
  let currentNode = root;
  for (const bit of encodedText) {
    if (bit === "0") {
      currentNode = currentNode.left;
    } else {
      currentNode = currentNode.right;
    }
    if (currentNode && currentNode.char) {
      decodedText += currentNode.char;
      currentNode = root;
    }
     else if (!currentNode) {
      return decodedText;
    }
  }
  return decodedText;
}

// 3. Huffman Coding with ES6 Map and Optimized Object Creation
// ----------------------------------------------------------------------

function buildHuffmanTreeES6(charFreq) {
  if (!charFreq || charFreq.length === 0) {
    return null;
  }
  const nodes = charFreq.map(([char, freq]) => ({ char, freq, left: null, right: null }));
  const priorityQueue = [...nodes];

  while (priorityQueue.length > 1) {
    priorityQueue.sort((a, b) => a.freq - b.freq);
    const left = priorityQueue.shift();
    const right = priorityQueue.shift();
    const newNode = {
      char: null,
      freq: left.freq + right.freq,
      left,
      right,
    };
    priorityQueue.push(newNode);
  }
  return priorityQueue[0];
}

function generateHuffmanCodesES6(node, code = "", codes = new Map()) {
  if (!node) return codes;
  if (node.char) {
    codes.set(node.char, code);
    return codes;
  }
  generateHuffmanCodesES6(node.left, code + "0", codes);
  generateHuffmanCodesES6(node.right, code + "1", codes);
  return codes;
}

function encodeTextES6(text, codes) {
  if (!codes) return "";
  let encodedText = "";
  for (const char of text) {
    encodedText += codes.get(char) || '';
  }
  return encodedText;
}

function decodeTextES6(encodedText, root) {
  if (!root) return "";
  let decodedText = "";
  let currentNode = root;
  for (const bit of encodedText) {
    currentNode = bit === "0" ? currentNode.left : currentNode.right;
    if (currentNode && currentNode.char) {
      decodedText += currentNode.char;
      currentNode = root;
    }
    else if (!currentNode) {
      return decodedText;
    }
  }
  return decodedText;
}

// 4. Huffman Coding with Class and Static Methods
// ------------------------------------------------------
class HuffmanCoder {
  constructor(charFreq) {
    this.charFreq = charFreq;
    this.root = this.buildHuffmanTree();
    this.codes = this.generateHuffmanCodes();
  }

  buildHuffmanTree() {
    if (!this.charFreq || this.charFreq.length === 0) {
      return null;
    }
    const nodes = this.charFreq.map(([char, freq]) => ({ char, freq, left: null, right: null }));
    const priorityQueue = [...nodes];

    while (priorityQueue.length > 1) {
      priorityQueue.sort((a, b) => a.freq - b.freq);
      const left = priorityQueue.shift();
      const right = priorityQueue.shift();
      const newNode = { char: null, freq: left.freq + right.freq, left, right };
      priorityQueue.push(newNode);
    }
    return priorityQueue[0];
  }

  generateHuffmanCodes(node = this.root, code = "", codes = {}) {
    if (!node) return codes;
    if (node.char) {
      codes[node.char] = code;
    } else {
      this.generateHuffmanCodes(node.left, code + "0", codes);
      this.generateHuffmanCodes(node.right, code + "1", codes);
    }
    return codes;
  }

  encodeText(text) {
    if (!this.codes) return "";
    let encodedText = "";
    for (const char of text) {
      encodedText += this.codes[char] || '';
    }
    return encodedText;
  }

  decodeText(encodedText) {
    if (!this.root) return "";
    let decodedText = "";
    let currentNode = this.root;
    for (const bit of encodedText) {
      currentNode = bit === "0" ? currentNode.left : currentNode.right;
      if (currentNode && currentNode.char) {
        decodedText += currentNode.char;
        currentNode = this.root;
      }
      else if (!currentNode) {
        return decodedText;
      }
    }
    return decodedText;
  }

  static getFrequencies(text) {
    const freqMap = new Map();
    for (const char of text) {
      freqMap.set(char, (freqMap.get(char) || 0) + 1);
    }
    return Array.from(freqMap.entries());
  }
}

// 5. Huffman Coding with Modern ES6 (Classes, Maps, Concise Syntax)
// ----------------------------------------------------------------------
class HuffmanEncoder {
  constructor(text) {
    this.text = text;
    this.frequencies = this.calculateFrequencies();
    this.root = this.buildTree();
    this.codes = this.generateCodes();
  }

  calculateFrequencies() {
    const freqMap = new Map();
    for (const char of this.text) {
      freqMap.set(char, (freqMap.get(char) || 0) + 1);
    }
    return Array.from(freqMap.entries());
  }

  buildTree() {
    if (!this.frequencies || this.frequencies.length === 0) {
      return null;
    }
    const nodes = Array.from(this.frequencies).map(([char, freq]) => ({ char, freq, left: null, right: null }));
    const queue = [...nodes];

    while (queue.length > 1) {
      queue.sort((a, b) => a.freq - b.freq);
      const left = queue.shift();
      const right = queue.shift();
      const parent = { char: null, freq: left.freq + right.freq, left, right };
      queue.push(parent);
    }
    return queue[0];
  }

  generateCodes(node = this.root, code = "") {
    const codes = new Map();
    const _generate = (node, code) => { // Inner recursive function
      if (!node) return;
      if (node.char) {
        codes.set(node.char, code);
      } else {
        _generate(node.left, code + "0");
        _generate(node.right, code + "1");
      }
    };
    _generate(node, code); // Start the recursion
    return codes;
  }

  encode() {
    if (!this.codes) return "";
    return this.text.split('').map(char => this.codes.get(char) || '').join('');
  }

  decode(encodedText) {
    if (!this.root) return "";
    let decodedText = '';
    let current = this.root;
    for (const bit of encodedText) {
      current = bit === '0' ? current.left : current.right;
      if (current && current.char) {
        decodedText += current.char;
        current = this.root;
      }
      else if (!current) {
        return decodedText;
      }
    }
    return decodedText;
  }
}

// Example Usage and Output
// ----------------------
const text = "hello world";
const charFreq = [...new Map().entries(new Map(text.split('').map(c => [c, text.split('').filter(x => x === c).length])))]; // Calculate character frequencies

console.log("Input Text:", text);
console.log("Character Frequencies:", charFreq);

// 1. Basic Huffman Coding
console.log("\n1. Basic Huffman Coding:");
const rootBasic = buildHuffmanTreeBasic(charFreq);
const codesBasic = generateHuffmanCodesBasic(rootBasic);
const encodedTextBasic = encodeTextBasic(text, codesBasic);
const decodedTextBasic = decodeTextBasic(encodedTextBasic, rootBasic);
console.log("Codes:", codesBasic);
console.log("Encoded Text:", encodedTextBasic);
console.log("Decoded Text:", decodedTextBasic);

// 2. Huffman Coding with Optimized Priority Queue (Min Heap)
console.log("\n2. Huffman Coding with Optimized Priority Queue (Min Heap):");
const rootOptimized = buildHuffmanTreeOptimized(charFreq);
const codesOptimized = generateHuffmanCodesOptimized(rootOptimized);
const encodedTextOptimized = encodeTextOptimized(text, codesOptimized);
const decodedTextOptimized = decodeTextOptimized(encodedTextOptimized, rootOptimized);
console.log("Codes:", codesOptimized);
console.log("Encoded Text:", encodedTextOptimized);
console.log("Decoded Text:", decodedTextOptimized);

// 3. Huffman Coding with ES6 Map
console.log("\n3. Huffman Coding with ES6 Map:");
const rootES6 = buildHuffmanTreeES6(charFreq);
const codesES6 = generateHuffmanCodesES6(rootES6);
const encodedTextES6 = encodeTextES6(text, codesES6);
const decodedTextES6 = decodeTextES6(encodedTextES6, rootES6);
console.log("Codes:", codesES6);
console.log("Encoded Text:", encodedTextES6);
console.log("Decoded Text:", decodedTextES6);

// 4. Huffman Coding with Class and Static Methods
console.log("\n4. Huffman Coding with Class and Static Methods:");
const coder = new HuffmanCoder(charFreq);
const encodedTextClass = coder.encodeText(text);
const decodedTextClass = coder.decodeText(encodedTextClass);
console.log("Codes:", coder.codes);
console.log("Encoded Text:", encodedTextClass);
console.log("Decoded Text:", decodedTextClass);

// 5. Huffman Coding with Modern ES6
console.log("\n5. Huffman Coding with Modern ES6:");
const encoder = new HuffmanEncoder(text);
const encodedTextModern = encoder.encode();
const decodedTextModern = encoder.decode(encodedTextModern);
console.log("Codes:", encoder.codes);
console.log("Encoded Text:", encodedTextModern);
console.log("Decoded Text:", decodedTextModern);
